## Basic Sorting Algorithms

<hr>

### What is Invariant?
- Invariant is a property that `remains unchanged`

    - At a particular program point, or throughout an algorithm, a function, a module 

__Invaritants__ can be used to `improve`/`optimize` an algorithm

<br>

### What is Incremental?
- An algorithm is incremental if it does ` NOT need to re-compute things after a small change (like adding a num to a sorted list)`

    - It can `reuse most of the work` already done to handle the change

<br>

### What is Stable?
- Maintains the relative order among elements

<img src="images/stable1.png" width="700px">
<img src="images/stable2.png" width="364px">

<br><hr>

### Lets look at 3 basic sorting algorithms:



<img src="images/sorts.png" width="500px">


<br><hr><br>

In [1]:
# TEST CASE
test0 = {
    'input': {
        'nums': [5, -12, 2, 6, 1, 23, 7, 7, -12, 6, 12, 1, -243, 1, 0]
    },
    'output': [-243, -12, -12, 0, 1, 1, 1, 2, 5, 6, 6, 7, 7, 12, 23]
}

<br><hr>

### Bubble Sort

#### Invariants:

> After every traversal, the list has the same elements

> In each traversal at most n-1 swaps are performed, where `n` is the length of the list

We can use the below variant to optimize the Bubble Sort
> After every traversal, the `largest` yet unsorted `element gets to its final place` (rightmost elements)


#### Incremental:

> If we append a number at the end of the list, it only take a __`few main-loop iterations`- Not very incremental__.

> However, if add anumber in the beginning of the list, it only takes __`1 main-loop iteration`- Very incremental__.


#### Stable:

> Bubble Sort is __stable__

> But careful, a small change (>= rather than > `strict conditional`) makes it non stable


In [2]:
def bubble_sort(nums):
    n = len(nums)

    # Iterate the outer loop n-1 times
    for x in range(n-1): 

        # Traverse through each element
        for i in range(n-1): 
            if nums[i] > nums[i+1]: # if curr el > next el
                nums[i], nums[i+1] = nums[i+1], nums[i] # swap them

    return nums

In [3]:
%%time
if bubble_sort(test0['input']['nums'][:]) == test0['output']:
    print('passed test0\n')

passed test0

CPU times: user 33 µs, sys: 5 µs, total: 38 µs
Wall time: 39.1 µs


<br><hr>

### Optimized Bubble Sort

#### Invariants:

> After every traversal, the `largest` unsorted `element gets to its final place` (rightmost elements)

How is using the above variant useful?

- Each iteration can avoid comparing the last element moved


#### Incremental:

> If we append a number at the end of the list, it only take a __`few main-loop iterations`- Not very incremental__.

> However, if add anumber in the beginning of the list, it only takes __`1 main-loop iteration`- Very incremental__.


#### Stable:

> Optimized Bubble Sort is __stable__

> But careful, a small change (>= rather than > `strict conditional`) makes it non stable


In [4]:
def bubble_sort_optimized(nums):
    n = len(nums)

    for x in range(n-1): # Iterate the outer loop n-1 times
       
        swapped = False # Reset swapped after completion of every inner loop

        # Decrementing range by 1 for every iteration done by the outer loop
        for i in range(n-1 - x): 
            if nums[i] > nums[i+1]:
                nums[i], nums[i+1] = nums[i+1], nums[i]
                swapped = True

        # if no swapping orccurs, the list or remain numbers in the list is already sorted, so return it
        if not swapped: 
            return nums
        
    return nums

In [5]:
%%time
if bubble_sort_optimized(test0['input']['nums'][:]) == test0['output']:
    print('passed test0\n')

passed test0

CPU times: user 30 µs, sys: 1 µs, total: 31 µs
Wall time: 34.1 µs


<br><hr>

### Selection Sort


#### Invariants:

> The swapped numbers on the left-most side will remain unchanged once swapped.


#### Incremental:

>  Selection Sort needs to do __all the work again__, so it is ___NOT___ Incremental


#### Stable:

> Selection Sort is __swapping non-consecutive elements__, so it is ___NOT___ __stable__.

> Ex: If there are 2 same elements, they will not be stable.


In [6]:
def get_min_idx(nums, cur_min_idx):

    # start the loop from cur_min_idx + 1 to the len(nums)-1
    for i in range(cur_min_idx+1, len(nums)):

        # if the i_th_el is less that cur_min_el
        if nums[i] < nums[cur_min_idx]:
            cur_min_idx = i # change the cur_min_idx to i

    return cur_min_idx


def selection_sort(nums):
    n = len(nums)

    # run the main-loop n-1 times
    for i in range(n-1):
        min_idx = get_min_idx(nums, i) # get the min_idx

        # swap the curr_el with the min_el
        nums[i], nums[min_idx] = nums[min_idx], nums[i]
    return nums

In [7]:
%%time
if selection_sort(test0['input']['nums'][:]) == test0['output']:
    print('passed test0\n')

passed test0

CPU times: user 35 µs, sys: 2 µs, total: 37 µs
Wall time: 41.2 µs


<br><hr>

### Insertion Sort

#### Invariants:

> Elements to the left are sorted and `grow by 1 each main-loop iteration`.

#### Incremental:

> Everything to the left is sorted but might not be in its final position.

> So, `it is Incremental` if w appending a new element at the end. 


#### Stable:

> Insertion Sort is __stable__.

> But careful, a small change (>= rather than > `strict conditional`) makes it non stable.

In [8]:
# Insertion Sort is like how we sort cards
# We assume that the firs element is sorted
# Then we take the card next to it out and see if we need to place the card before the sorted card or after
# We repeat this process until, untill all the cards are sorted

def insertion_sort(nums):
    n = len(nums)
    
    # assuming the first element (at idx=0) is sorted, 
    # so the main loop need to start at idx=1 to get the next el/card which is not sorted
    for x in range(1, n): # running n-1 times

        cur_el = nums.pop(x) # take the curr_el/card out which is next to the sorted_el/card

        # i=x-1 because we will need to start checking from the sorted_el which comes right before the cur_el
        i = x - 1 

        # the loop has a condition which helps find the idx where the cur_el needs to be placed
        while i >= 0 and nums[i] > cur_el:
            i -= 1
        nums.insert(i+1, cur_el) # place that cur_el/card at the correct spot
    
    return nums

In [9]:
%%time
if insertion_sort(test0['input']['nums'][:]) == test0['output']:
    print('passed test0\n')

passed test0

CPU times: user 25 µs, sys: 5 µs, total: 30 µs
Wall time: 32.9 µs


<br><hr><br>

<div style="display:flex; justify-content: center;">
<table class="tg" style="align-self: center;">
<thead>
  <tr>
    <th class="tg-bzci" colspan="5"><span style="font-weight:bold;font-style:italic; font-size:28px;"><center>Sorting Complexities</center></span></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-eroe" colspan="2"><span style="font-weight:bold">Bubble Sort</span></td>
    <td class="tg-0lax" rowspan="5"></td>
    <td class="tg-dffh" colspan="2"><span style="font-weight:bold">Optimized Bubble Sort</span></td>
  </tr>
  <tr>
    <td class="tg-1wig">Time Complexity (wort-case):</td>
    <td class="tg-rvyq">O(2^n)</td>
    <td class="tg-fymr">Time Complexity (wort-case):</td>
    <td class="tg-rvyq"><span>O(2^n)</span></td>
  </tr>
  <tr>
    <td class="tg-1wig">Time Complexity (average-case):</td>
    <td class="tg-rvyq">O(2^n)</td>
    <td class="tg-fymr">Time Complexity (average-case):</td>
    <td class="tg-rvyq"><span>O(2^n)</span></td>
  </tr>
  <tr>
    <td class="tg-1wig">Time Complexity (best-case):</td>
    <td class="tg-rvyq">O(2^n)</td>
    <td class="tg-fymr">Time Complexity (best-case):</td>
    <td class="tg-rvyq"><span>O(n)</span></td>
  </tr>
  <tr>
    <td class="tg-1wig">Space Complexity:</td>
    <td class="tg-ihln">O(1)</td>
    <td class="tg-1wig">Space Complexity:</td>
    <td class="tg-ihln"><span>O(1)</span></td>
  </tr>
  <tr>
    <td class="tg-0lax" colspan="2"></td>
    <td class="tg-0lax"></td>
    <td class="tg-0lax" colspan="2"></td>
  </tr>
  <tr>
    <td class="tg-eroe" colspan="2"><span style="font-weight:bold">Selection Sort</span></td>
    <td class="tg-0lax" rowspan="5"></td>
    <td class="tg-eroe" colspan="2"><span style="font-weight:bold">Insertion Sort</span></td>
  </tr>
  <tr>
    <td class="tg-1wig">Time Complexity (wort-case):</td>
    <td class="tg-ihln">O(2^n)</td>
    <td class="tg-1wig">Time Complexity (wort-case):</td>
    <td class="tg-ihln">O(2^n)</td>
  </tr>
  <tr>
    <td class="tg-1wig">Time Complexity (average-case):</td>
    <td class="tg-ihln">O(2^n)</td>
    <td class="tg-1wig">Time Complexity (average-case):</td>
    <td class="tg-ihln">O(2^n)</td>
  </tr>
  <tr>
    <td class="tg-1wig">Time Complexity (best-case):</td>
    <td class="tg-ihln">O(2^n)</td>
    <td class="tg-1wig">Time Complexity (best-case):</td>
    <td class="tg-ihln">O(2^n)</td>
  </tr>
  <tr>
    <td class="tg-1wig">Space Complexity:</td>
    <td class="tg-ihln">O(1)</td>
    <td class="tg-1wig">Space Complexity:</td>
    <td class="tg-ihln">O(1)</td>
  </tr>
</tbody>
</table>
</div>
<br><hr>

### Lets test each Sortin Algorithm using a long random list

In [10]:
import random
in_list = list(range(10000))
out_list = list(range(10000))
random.shuffle(in_list)
big_test = {
    'input': {
        'nums': in_list
    },
    'output': out_list
}

In [11]:
%%time
if bubble_sort(big_test['input']['nums'][:]) == big_test['output']:
    print('BUBBLE SORT\n')

BUBBLE SORT

CPU times: user 6.09 s, sys: 57 ms, total: 6.15 s
Wall time: 6.15 s


In [12]:
%%time
if bubble_sort_optimized(big_test['input']['nums'][:]) == big_test['output']:
    print('OPTIMIZED BUBBLE SORT\n')

OPTIMIZED BUBBLE SORT

CPU times: user 3.99 s, sys: 31.2 ms, total: 4.02 s
Wall time: 4.03 s


In [13]:
%%time
if selection_sort(big_test['input']['nums'][:]) == big_test['output']:
    print('SELECTION SORT\n')

SELECTION SORT

CPU times: user 1.59 s, sys: 13.8 ms, total: 1.61 s
Wall time: 1.6 s


In [14]:
%%time
if insertion_sort(big_test['input']['nums'][:]) == big_test['output']:
    print('INSERTION SORT\n')

INSERTION SORT

CPU times: user 946 ms, sys: 6.93 ms, total: 953 ms
Wall time: 952 ms


<br><hr>

### Ranking of each Sorting Algorithms for a `long-random list of numbers`:

#### 1. Insertion Sort (952 ms)
#### 2. Selection Sort (1.60 s)
#### 3. Optimized Bubbke Sort (4.03 s)
#### 4. Bubble Sort (6.15 s)